BinaryEncoding.md

```diff
@@ -3,120 +3,143 @@
+
+### Subtree deduplication / nullary macros
+Identifies and prunes structurally identical nodes and trees of nodes. Most applications contain significant amounts of structural 
+duplication that is not completely erased by generic compression.
+**Non-nullary macros** are an extension of this technique that enables further compression at the cost of additional complexity.
+
+### Index tables
+Modules contain multiple index tables that assign indexes to key pieces of information like opcodes or data types. This enables
+compatibility between implementations and allows information to be represented more efficiently.
+
+### Sections
+Modules are split up into sections with well-defined contents that can refer to each other and are identified by name.
+The use of names allows new section types to be introduced in the future.
+
+### Strings
+Strings are encoded as null-terminated [UTF8](http://unicode.org/faq/utf_bom.html#UTF8).
+
+# v8-native module structure
+
+The following documents the current v8-native prototype format, not the binary encoding intended for standardization.
+
+## High-level structure
+A module contains (in this order):
+* A stream of sections, containing for each section:
+  - ```uint8```: A [section type identifier](https://github.com/v8/v8/blob/master/src/wasm/wasm-module.h#L26) for the section
+  - The section body (defined below by section type)
+
+### Memory section
+* ```uint8```: The minimum size of the module heap in bytes, as a power of two
+* ```uint8```: The maximum size of the module heap in bytes, as a power of two
+* ```uint8```: ```1``` if the module's memory is externally visible
+
+### Signatures section
+* [```varuint32```](#varuint32): The number of function signatures in the section
+* For each function signature:
+  - ```uint8```: The number of parameters
+  - ```uint8```: The function return type, as a [LocalType](https://github.com/v8/v8/blob/master/src/wasm/wasm-opcodes.h#L16)
+  - For each parameter:
+    + ```uint8```: The parameter type, as a LocalType
+
+### Functions section
+This section must be preceded by a [Signatures](#signatures-section) section.
+
+* ```varuint32```: The number of functions in the section
+* For each function:
+  - ```uint8```: The [function declaration bits](https://github.com/v8/v8/blob/master/src/wasm/wasm-module.h#L39)
+  - ```uint16```: The function signature (as an index into the Signatures section)
+  - If the ```kDeclFunctionName``` bit is set:
+    + ```uint32```: The offset of the function name in the file.
+  - If the ```kDeclFunctionImport``` bit is set, **the function entry ends here**
+  - If the ```kDeclFunctionLocals``` bit is set:
+    + ```uint16```: The number of i32 locals
+    + ```uint16```: The number of i64 locals
+    + ```uint16```: The number of f32 locals
+    + ```uint16```: The number of f64 locals
+  - ```uint16```: The size of the function body, in bytes
+  - The function body
+
+### Globals section
```

Comment:
That's not possible, because the length header is # of globals instead of # of bytes. If we properly laid out sections with length-in-bytes headers we could reserve section types for internal use.